subsys/ipc/ipc_service/lib/Kconfig.icmsg

```diff
@@ -27,15 +27,53 @@ config IPC_SERVICE_ICMSG_SHMEM_ACCESS_TO_MS
 	  Maximum time to wait, in milliseconds, for access to send data with
 	  backends basing on icmsg library. This time should be relatively low.
 
-# The Icmsg library in its simplicity requires the system workqueue to execute
-# at a cooperative priority.
-config SYSTEM_WORKQUEUE_PRIORITY
-	range -256 -1
-
 config IPC_SERVICE_ICMSG_BOND_NOTIFY_REPEAT_TO_MS
 	int "Bond notification timeout in miliseconds"
 	range 1 100
 	default 1
 	help
 	  Time to wait for remote bonding notification before the
 	  notification is repeated.
+
+config IPC_SERVICE_BACKED_ICMSQ_WQ_ENABLE
+	bool "Use dedicated workqueue"
+	default y
+	help
+	  Enable dedicated workqueue thread for the ICMsg backend.
+	  Disabling this configuration will cause the ICMsg backend to
+	  process incoming data through the system workqueue context, and
+	  therefore reduces the RAM footprint of the backend.
+	  Disabling this config may result in deadlocks in certain usage
+	  scenarios, such as when synchronous IPC is executed from the system
+	  workqueue context.
+	  The callbacks coming from the backend are executed from the workqueue
+	  context.
+	  When the option is disabled, the user must obey the restrictions
+	  imposed by the system workqueue, such as never performing blocking
+	  operations from within the callback.
+
+if IPC_SERVICE_BACKED_ICMSQ_WQ_ENABLE
+
+config IPC_SERVICE_BACKEND_ICMSG_WQ_STACK_SIZE
+	int "Size of RX work queue stack"
+	default 1024
```

Comment:
Leaving dedicated WQ enabled by default is risky - from the top of my mind some of the applications are increasing system workque stack to fit. Now such configurations may fail.

Reply:
I don't know if I understand the problem. The defconfig for a particular platform/CPU should set the stack size to be large enough to strike the balance between the most common use-cases and memory utilization.